java/src/main/native/src/row_conversion.cu

```diff
@@ -923,106 +910,89 @@ __global__ void copy_from_rows(const size_type num_rows, const size_type num_col
+        dst[b] = src[b];
       }
+#endif
     }
+  }
+
+  {
+    auto const tile = tile_infos[blockIdx.x];
+    auto const rows_in_tile = tile.num_rows();
+    auto const cols_in_tile = tile.num_cols();
+    auto const tile_row_size = tile.get_shared_row_size(col_offsets, col_sizes);
 
 #ifdef ASYNC_MEMCPY_SUPPORTED
-    auto &processing_barrier = tile_barrier[processing_index % NUM_TILES_PER_KERNEL_LOADED];
     // ensure our data is ready
-    processing_barrier.arrive_and_wait();
+    tile_barrier.arrive_and_wait();
 #else
     group.sync();
 #endif // ASYNC_MEMCPY_SUPPORTED
 
-    auto const tile = tile_infos[blockIdx.x * NUM_TILES_PER_KERNEL_FROM_ROWS + processing_index];
-    auto const rows_in_tile = tile.num_rows();
-    auto const cols_in_tile = tile.num_cols();
-    auto const tile_row_size = tile.get_shared_row_size(col_offsets, col_sizes);
-
     // now we copy from shared memory to final destination.
     // the data is laid out in rows in shared memory, so the reads
     // for a column will be "vertical". Because of this and the different
     // sizes for each column, this portion is handled on row/column basis.
     // to prevent each thread working on a single row and also to ensure
     // that all threads can do work in the case of more threads than rows,
     // we do a global index instead of a double for loop with col/row.
-    for (int index = threadIdx.x; index < rows_in_tile * cols_in_tile; index += blockDim.x) {
-      auto const relative_col = index % cols_in_tile;
-      auto const relative_row = index / cols_in_tile;
-      auto const absolute_col = relative_col + tile.start_col;
-      auto const absolute_row = relative_row + tile.start_row;
+    for (int relative_row = warp.thread_rank(); relative_row < rows_in_tile;
+         relative_row += warp.size()) {
 
+      auto const absolute_row = relative_row + tile.start_row;
       auto const shared_memory_row_offset = tile_row_size * relative_row;
-      auto const shared_memory_offset =
-          col_offsets[absolute_col] - col_offsets[tile.start_col] + shared_memory_row_offset;
-      auto const column_size = col_sizes[absolute_col];
 
-      int8_t *shmem_src = &shared[processing_index % stages_count][shared_memory_offset];
-      int8_t *dst = &output_data[absolute_col][absolute_row * column_size];
+      for (int relative_col = warp.meta_group_rank(); relative_col < cols_in_tile;
+           relative_col += warp.meta_group_size()) {
+        auto const absolute_col = relative_col + tile.start_col;
+
+        auto const shared_memory_offset =
+            col_offsets[absolute_col] - col_offsets[tile.start_col] + shared_memory_row_offset;
+        auto const column_size = col_sizes[absolute_col];
+
+        int8_t *shmem_src = &shared[shared_memory_offset];
+        int8_t *dst = &output_data[absolute_col][absolute_row * column_size];
 
-      MEMCPY(dst, shmem_src, column_size, processing_barrier);
+        MEMCPY(dst, shmem_src, column_size, tile_barrier);
```

Comment:
I usually see the `memcpy` APIs operating on `void*` pointers. Using `int8_t` is of course still valid, but I'm not sure if there will be any conversion to `void*` under the hood? If so, the compiler may issue warnings.

Reply:
Pardon me, but `T*` to `void*` [is a ~standard~ *implicit* conversion](https://en.cppreference.com/w/cpp/language/implicit_conversion). Why would the compiler complain?

Reply:
They may complain about some implicit conversions. I'm not sure if the compilers will complain about this case.

Reply:
I'd recommend an explicit cast only if the compiler complains. (And that would be surprising.)